daemon/delete.go

```diff
@@ -92,13 +95,29 @@ func (daemon *Daemon) Rm(container *Container) error {
 		return fmt.Errorf("Container %v not found - maybe it was already destroyed?", container.ID)
 	}
 
+	// Container state RemovalInProgress should be used to avoid races.
+	if err := container.SetRemovalInProgress(); err != nil {
+		return fmt.Errorf("Failed to set container state to RemovalInProgress: %s", err)
+	}
+
+	defer container.ResetRemovalInProgress()
+
 	if err := container.Stop(3); err != nil {
 		return err
 	}
 
-	// Deregister the container before removing its directory, to avoid race conditions
-	daemon.idIndex.Delete(container.ID)
-	daemon.containers.Delete(container.ID)
+	// Mark container dead. We don't want anybody to be restarting it.
+	container.SetDead()
+
+	// If force removal is required, delete container from various
+	// indexes even if removal failed.
+	defer func() {
+		if err != nil && forceRemove {
```

Comment:
I tested it and err is not nil.

I have defined a named return error as follows.

func (daemon *Daemon) Rm(container *Container, forceRemove bool) (err error)

So this err inside defer must be referring to that. And while I am returning, I am always explicitly returning an error object. So "err" ends up referring to that. 

But if it is confusing and if you have a preferred way to handle it I can do that. I guess I can stop redeclaring err in various calls and that might help.